athena audit logs - query rate limiter

tobiaszheller opened the pull request:
Part of https://github.com/gravitational/teleport.e/issues/894
RFD: https://github.com/gravitational/teleport/blob/master/rfd/0118-scalable-audit-logs.md

This PR adds rate limit feature to athena audit logs on searching events. 
Rate limit is needed to allow throttling of tenants in Cloud.

smallinsky commented on lib/events/athena/querier.go:
```diff
@@ -133,12 +157,33 @@ func (q *querier) SearchSessionEvents(fromUTC, toUTC time.Time, limit int,
+}
+
+func (q *querier) searchEvents(ctx context.Context, req searchEventsRequest) ([]apievents.AuditEvent, string, error) {
+	if q.limiter != nil && !q.limiter.Allow() {
+		// TODO(tobiaszheller): return how much use should sleep at some point.
+		return nil, "", trace.LimitExceeded("rate limit exceeded for searching events")
+	}
```
Just want to ensure that we want to rate limit internal searchEvents call instead of eternal Athena API calls.

I'm thinking about the case where `searchEvents` function is invoked for a large sets of `[start, end]` and internally Athena API can be called multiple times by  searchEvents->fetchResults->GetQueryResults within the single `searchEvents` execution.

tobiaszheller replied:
yes, this is fine. Typically the getResults call has much larger quota: https://docs.aws.amazon.com/athena/latest/ug/service-limits.html#service-limits-api-calls
so we should be fine limiting it on searchEvents endpoint.

smallinsky commented on lib/events/athena/querier.go:
```diff
@@ -133,12 +157,33 @@ func (q *querier) SearchSessionEvents(fromUTC, toUTC time.Time, limit int,
+func (q *querier) searchEvents(ctx context.Context, req searchEventsRequest) ([]apievents.AuditEvent, string, error) {
+	if q.limiter != nil && !q.limiter.Allow() {
+		// TODO(tobiaszheller): return how much use should sleep at some point.
+		return nil, "", trace.LimitExceeded("rate limit exceeded for searching events")
```
nit:
We can consider decoupling SearchEventsLimiter from AtenaAuditLoger: 
```
type SearchEventsLimmiter struct {
  limiter  
  inner events.AuditLogger
}

func  .. EmitAuditEvent(context.Context,  ...) ... { 
   return inner.EmitAuditEvent(...)
}
func  .. SearchSessionEvents(context.Context, ...) ... { 
      if limit {
       return ...
    }
    return inner.SearchSessionEvents(..)
}
func  .. SearchEvents(context.Context, ...) ... { 
      if limit {
       return ...
    }
    return inner.SearchEvents(..)
}
```

tobiaszheller replied:
reworked in https://github.com/gravitational/teleport/pull/24918/commits/dfc304cf14978115ef6f44605b868b9752047c71

tobiaszheller commented on lib/events/athena/querier.go:
```diff
@@ -69,6 +74,11 @@ type querierConfig struct {
 
+	// limiterRate defines rate at which search_event rate limiter is filled (optional).
+	limiterRate float64
+	// limiterBurst defines rate limit bucket capacity (optional).
+	limiterBurst int
```
This is already checked in package top level config: https://github.com/gravitational/teleport/blob/bdc3db931ef62b9d383a15023ebb0125b467cb40/lib/events/athena/athena.go#L189-L202
I didn't want to duplicate that code, let me know if this is fine or not.

tobiaszheller commented on lib/events/athena/querier.go:
```diff
@@ -69,6 +74,11 @@ type querierConfig struct {
 	getQueryResultsInterval time.Duration
 
+	// limiterRate defines rate at which search_event rate limiter is filled (optional).
+	limiterRate float64
```
good catch, fixed in https://github.com/gravitational/teleport/pull/24918/commits/5522a39efdff698b45046b34849d14c543922b6b

tobiaszheller commented on lib/events/search_limiter.go:
```diff
@@ -0,0 +1,86 @@
+	if cfg.LimiterBurst <= 0 {
+		return trace.BadParameter("limiterBurst cannot be less or equal to 0")
+	}
+	if math.Abs(cfg.LimiterRate) < 1e-9 {
```
sorry for missing docs for that config. `LimiterRate` defines how many tokens and added per seconds. 
I am checking it with `1e-9` just to have very small delta when comparing float against 0. On one hand it should be equal to 0 because if it's not present in url params, it will be default, so 0. @rosstimothy do you think I should just better document it? 

@smallinsky i don't think you can define rps using uint, someone may want to set it 0.5 for example. And defining it in miliseconds will be even more confusing. 

One thing that we can do is use helper fn from `rate` package [Every](https://pkg.go.dev/golang.org/x/time/rate#Every). You define interval (time.Duration) at which token is filled. Do you think it will be easier to understand?

smallinsky replied:
0.5 RPS or 10.1 RPS values as rate limit should be quite rate but I don't have strong about it so please free to ignore my comment.

> I don't thing that we can do is use helper fn from rate package [Every](https://pkg.go.dev/golang.org/x/time/rate#Every). You define interval (time.Duration) at which token is filled. Do you think it will be easier to understand?

I think that current behaviour with 1s time unit is fine and readable. It should be quite easy to extend in the future but if we don't have case for Athena we should skip it for now and keep it simple.

tobiaszheller commented on lib/events/search_limiter_test.go:
```diff
@@ -0,0 +1,133 @@
+	"github.com/gravitational/teleport/lib/events"
+)
+
+func TestSearchEventsLimiter(t *testing.T) {
```
solved in https://github.com/gravitational/teleport/pull/24918/commits/4fa09ecfb50fe613c8bb0ee5c4eb995e6d627f37

tobiaszheller commented:
@rosstimothy  https://github.com/gravitational/teleport/pull/24918/commits/4fa09ecfb50fe613c8bb0ee5c4eb995e6d627f37
addresses all comments from PR. I think only https://github.com/gravitational/teleport/pull/24918#discussion_r1175287698 remains open and would like to hear your opinions.

tobiaszheller commented:
@smallinsky @rosstimothy in https://github.com/gravitational/teleport/pull/24918/commits/26b6d79ff04af347caca9ed98dc04ec40e7b8580
I have reworked config, it accepts now 3 params:
```
	// RefillTime is amount of time between refills.
	RefillTime time.Duration
	// RefillAmount is the number of tokens that are added to the bucket during a refill.
	RefillAmount int
	// Burst defines number of available tokens. It's initially full and refilled
	// based on refillAmount and refillDuration.
	Burst int
```
similary to how our per IP limiter does: https://github.com/gravitational/oxy/blob/c59990dc8c641ec79a7f6eee48a20c7e8f4c69c6/ratelimit/tokenlimiter.go#L31, I just used more descriptive names.

Let me know what do you think?

tobiaszheller commented on lib/events/search_limiter.go:
```diff
@@ -0,0 +1,94 @@
+
+// SearchEventsLimiterConfig is configuration for SearchEventsLimiter.
+type SearchEventsLimiterConfig struct {
+	// RefillTime is amount of time between refills.
```
addresses in https://github.com/gravitational/teleport/pull/24918/commits/788efc45af9b0f6cb4f262799b1deb021b074e47

rosstimothy reviewed (commented):
LGTM but seems like there is now a broken test case:

```
=== RUN   TestAthenaAuditLogSetup/config_with_rate_limit_-_should_use_events.SearchEventsLimiter
{"caller":"athena/consumer.go:506","component":"athena","error":"operation error SQS: ReceiveMessage, failed to resolve service endpoint, an AWS region is required, but was not found","level":"error","message":"Failure processing SQS messages","timestamp":"2023-04-27T10:16:33Z"}
    service_test.go:322: 
        	Error Trace:	/__w/teleport/teleport/lib/service/service_test.go:322
        	            				/__w/teleport/teleport/lib/service/service_test.go:338
        	Error:      	Received unexpected error:
        	            	LimiterRefillAmount must be greater than 0 if LimiterBurst is used
        	Test:       	TestAthenaAuditLogSetup/config_with_rate_limit_-_should_use_events.SearchEventsLimiter
{"caller":"athena/consumer.go:506","component":"athena","error":"operation error SQS: ReceiveMessage, failed to resolve service endpoint, an AWS region is required, but was not found","level":"error","message":"Failure processing SQS messages","timestamp":"2023-04-27T10:16:33Z"}
--- FAIL: TestAthenaAuditLogSetup/config_with_rate_limit_-_should_use_events.SearchEventsLimiter (0.00s)
```

tobiaszheller commented:
> LGTM but seems like there is now a broken test case:
> 
> ```
> === RUN   TestAthenaAuditLogSetup/config_with_rate_limit_-_should_use_events.SearchEventsLimiter
> {"caller":"athena/consumer.go:506","component":"athena","error":"operation error SQS: ReceiveMessage, failed to resolve service endpoint, an AWS region is required, but was not found","level":"error","message":"Failure processing SQS messages","timestamp":"2023-04-27T10:16:33Z"}
>     service_test.go:322: 
>         	Error Trace:	/__w/teleport/teleport/lib/service/service_test.go:322
>         	            				/__w/teleport/teleport/lib/service/service_test.go:338
>         	Error:      	Received unexpected error:
>         	            	LimiterRefillAmount must be greater than 0 if LimiterBurst is used
>         	Test:       	TestAthenaAuditLogSetup/config_with_rate_limit_-_should_use_events.SearchEventsLimiter
> {"caller":"athena/consumer.go:506","component":"athena","error":"operation error SQS: ReceiveMessage, failed to resolve service endpoint, an AWS region is required, but was not found","level":"error","message":"Failure processing SQS messages","timestamp":"2023-04-27T10:16:33Z"}
> --- FAIL: TestAthenaAuditLogSetup/config_with_rate_limit_-_should_use_events.SearchEventsLimiter (0.00s)
> ```

fixed in https://github.com/gravitational/teleport/pull/24918/commits/d7e491e943caab1c7fed5c5437a9762c4d719ff5

Commit:
athena audit logs - query rate limiter
